src/Ormolu/Utils/Fixity.hs

```diff
@@ -0,0 +1,63 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module Ormolu.Utils.Fixity
+  ( getFixityOverridesForSourceFile,
+    parseFixityDeclarationStr,
+  )
+where
+
+import Control.Exception (throwIO)
+import Control.Monad.IO.Class
+import Data.Bifunctor (first)
+import Data.IORef
+import qualified Data.Map.Strict as Map
+import qualified Data.Text as T
+import qualified Data.Text.IO as T
+import Ormolu.Exception
+import Ormolu.Fixity
+import Ormolu.Fixity.Parser
+import Ormolu.Utils.Cabal
+import System.Directory
+import System.FilePath
+import Text.Megaparsec (errorBundlePretty)
+
+-- | Attempt to locate and parse a @.ormolu@ file. If it does not exist,
+-- empty fixity map is returned. This function maintains a cache of fixity
+-- overrides where cabal file paths act as keys.
+getFixityOverridesForSourceFile ::
+  MonadIO m =>
+  -- | 'CabalInfo' already obtained for this source file
+  CabalInfo ->
+  m FixityMap
+getFixityOverridesForSourceFile CabalInfo {..} = liftIO $ do
+  cacheRef <- newIORef Map.empty
+  case ciCabalFilePath of
+    Nothing -> return Map.empty
+    Just cabalPath -> do
+      cache <- readIORef cacheRef
+      case Map.lookup cabalPath cache of
+        Nothing -> do
+          let dotOrmolu = replaceFileName cabalPath ".ormolu"
+          exists <- doesFileExist dotOrmolu
+          if exists
+            then do
+              dotOrmoluRelative <- makeRelativeToCurrentDirectory dotOrmolu
+              contents <- T.readFile dotOrmolu
+              case parseFixityMap dotOrmoluRelative contents of
+                Left errorBundle ->
+                  throwIO (OrmoluFixityOverridesParseError errorBundle)
+                Right x -> do
+                  modifyIORef cacheRef (Map.insert cabalPath x)
```

Comment:
```suggestion
                  modifyIORef' cacheRef (Map.insert cabalPath x)
```

Reply:
I'm not sure we are gaining here anything from strictness, but it doesn't hurt either!

Reply:
In contrived cases (lots of files, each with their own `.ormolu` file, so cache entries are never read), there could be a thunk buildup in `cacheRef`.